### ISSUE TITLE
question about PDF structure, page elimination, metadata, and quality

### ISSUE THREAD

**Issue by piramiday** [qpdf/qpdf#447]
first of all, thanks for the awesome project! :+1:
it took me a while to find you, but QPDF should definitely be suggested as a first choice in most PDF handling questions in linux forums! :1st_place_medal: 

I'd like to ask a question about PDF structure, page elimination, metadata, and quality.
I have a PDF, say `old.pdf` (268 pages, 37,277 kB), from which I'd like to remove a couple of pages.

_NITROPDF_
for starters, I used a commercial program like NitroPDF to do that.
the new file, say `new.pdf`, _effectively_ does not contain the page I deleted (266 pages) but it's slightly bigger in filesize (37,279 kB, thus +2 kB).
out of curiosity, I compared the files through a hex dump, and indeed they are **byte for byte identical until the very end**, where the NitroPDF-modified file has some metadata appended:

```
$ diff old.pdf.txt new.pdf.txt
2383867,2383868c2383867,2383954
< 02467590  25 25 45 4f 46 0a                                 |%%EOF.|
< 02467596
---
> 02467590  25 25 45 4f 46 0a 31 20  30 20 6f 62 6a 0d 0a 3c  |%%EOF.1 0 obj..<|
> 024675a0  3c 2f 54 79 70 65 20 2f  43 61 74 61 6c 6f 67 0d  |</Type /Catalog.|
> 024675b0  0a 2f 50 61 67 65 73 20  32 20 30 20 52 0d 0a 2f  |./Pages 2 0 R../|
> 024675c0  4d 65 74 61 64 61 74 61  20 31 35 31 34 36 20 30  |Metadata 15146 0|
> 024675d0  20 52 0d 0a 3e 3e 0d 0a  65 6e 64 6f 62 6a 0d 0a  | R..>>..endobj..|
> 024675e0  32 20 30 20 6f 62 6a 0d  0a 3c 3c 2f 54 79 70 65  |2 0 obj..<</Type|
[snip]
> 02467ad0  0a 3e 3e 0d 0a 73 74 61  72 74 78 72 65 66 0d 0a  |.>>..startxref..|
> 02467ae0  33 38 31 37 33 31 33 31  0d 0a 25 25 45 4f 46 0d  |38173131..%%EOF.|
> 02467af0  0a                                                |.|
> 02467af1
```
I guess that **`new.pdf` still contains the removed pages, which got hidden through metadata**.
the quality is identical.

_QPDF_
right after that test I discovered QPDF and did the same page elimination.
the resulting file, say `q.pdf`, is 266 pages but this time it's quite smaller, 34,957 kB, thus -2 MB.
it's entirely possible that the two pages, being images, were roughly 1 MB each.
this time, though, **the two files are almost completely different**.
how can I be confident that the new PDF contains the same data of the original one, minus those two undesired pages?
I would have expected some different metadata and different "blocks" here and there, but from a preliminary layman comparison it seems the files have very little in common.

how does qpdf eliminate pages and recreate the PDF?
in other words, **is there a good way to compare the bowels of these various PDF files, in order to see with my own eyes what changed**?
I hope I've made my question clear. thanks!

**Comment by dreua** [comment]
I'm not an expert here but I'm curious if qpdf's qdf mode could help in this case. Try converting both pdfs to qdf pdfs by `qpdf --qdf your-infile.pdf your-outfile.pdf` and compare the results.

*From the [manual](http://qpdf.sourceforge.net/files/qpdf-manual.pdf):*

> The purpose of QDF form is to make it possible to edit PDF files, with some restrictions, in an ordinary text editor. This can be very useful for experimenting with different PDF constructs or formaking one-off edits to PDF files (though there are other reasons why this may not always work).

**Comment by piramiday** [comment]
I'm not familiar with the `--qdf` format, but that command just **doubled the filesize** of each document.
anyway, they still look fully different from a rough hexdump byte-by-byte comparison.

**Comment by jberkenbilt** [comment]
qdf mode is not going to help you here, and what you're describing is completely normal. PDF files are basically big, cross-referenced data structures that encode file structure and content, and qpdf's main gig is rearranging the files in different ways without changing the content. Doing byte-wise comparison of PDF files is almost never a useful exercise, though qpdf is itself mostly idempotent when run with the same arguments, meaning running qpdf on its own output usually creates a very similar file. I have high confidence that qpdf is not going to damage the content of your file if it exits normally. By my recollection, in qpdf's 12 year history, there has been only one bug that resulted in qpdf losing information on a normal run, and that bug was caught almost immediately after being introduced and was not part of any qpdf version that made it into any of the standard Linux distributions. That said, if you have any doubts, you can use ghostscript or some other tool that can render the PDF files as images and compare the images. You should then throw away the images as they are just used for comparison. If you do this, you must use a non-lossy format like png or tiff, not jpeg. In qpdf's test suite, I use ghostscript to convert PDF files to multi-page tiff files and then use tiffcmp to compare them. This is how I validate that qpdf is not damaging the content. Note that qdf mode almost always results in much larger files because, where possible, it removes compression so that files can be hand-inspected. qdf mode is not intended as a final format for any file. There is more discussion of this in the docs.

Hopefully this helps. I can provide detailed ghostscript commands, or you might be able to use ImageMagick, which calls ghostscript internally when you convert a PDF to an image, but maybe it's not necessary.

**Comment by jberkenbilt** [comment]
Regarding NitroPDF, I'm not familiar with the tool, but the PDF format does support modifying a file by appending. Basically you append something to the end that says to replace certain objects inside a file with a new version. This is useful in some cases, but obviously not what you want if you are removing a page and don't want to be able to recover the original. qpdf is not capable of writing files in this fashion (as I don't think this is useful for what qpdf is intended to do), but it has no problem reading them.

I'm glossing over the details here. If you want more depth, let me know. Otherwise, consider this reassurance that what you are seeing is totally normal, expected behavior.

**Comment by piramiday** [comment]
first of all, thanks for the stream-of-consciousness reply, @jberkenbilt! :zany_face:
you really answered lots of my questions.


> Doing byte-wise comparison of PDF files is almost never a useful exercise

yeah, I know. :disappointed: 
if I was really interested in doing that, though, **could you think of a way to check whether a PDF contains a blob of binary, say, an image?**


> In qpdf's test suite, I use ghostscript to convert PDF files to multi-page tiff files and then use tiffcmp to compare them.

my main point was to avoid losing image quality, so I'll definitely check your commands out!
it seems you are creating a single TIFF image from a PDF file like so:
https://github.com/qpdf/qpdf/blob/4977a7efa5d0b2d10f57af646e54a91d94d088ab/qpdf/qtest/qpdf.test#L4637-L4638
and then you are comparing the two:
https://github.com/qpdf/qpdf/blob/4977a7efa5d0b2d10f57af646e54a91d94d088ab/qpdf/qtest/qpdf.test#L4645
I have tried and it seems that `diff`ing `old.pdf.tif` with both `new.pdf.tif` and `q.pdf.tif` yields a return code of 1, since there is a difference of 2 pages, while a `diff` of `new.pdf.tif` and `q.pdf.tif` yields a 0!
I guess that means that the ghostscript conversion produced two identical bit-by-bit images?
final follow up question: **why aren't the two files identical, then?**
I can diff them using something like `vbindiff`, and I get lots of differences like this:
```
q.pdf.tif                                                                        
[snip]
0000 0120: 66 08 10 00 47 50 4C 20  47 68 6F 73 74 73 63 72  f...GPL  Ghostscr  
0000 0130: 69 70 74 20 39 2E 32 36  00 00 32 30 32 30 3A 30  ipt 9.26 ..2020:0  
0000 0140: 36 3A 31 38 20 30 39 3A  30 34 3A 31 33 00 00 00  6:18 09: 04:13...  
[snip]

new.pdf.tif                                                                        
[snip]
0000 0120: 66 08 10 00 47 50 4C 20  47 68 6F 73 74 73 63 72  f...GPL  Ghostscr  
0000 0130: 69 70 74 20 39 2E 32 36  00 00 32 30 32 30 3A 30  ipt 9.26 ..2020:0  
0000 0140: 36 3A 31 38 20 30 39 3A  30 34 3A 30 30 00 00 00  6:18 09: 04:00...  
[snip]
```
where the last bytes before the snip have a "04:00" in `new.pdf.tif` and "04:13" in `q.pdf.tif`.
I know it does not make sense to compare files like this, but I'd like to understand what's going on, if possible.

**Comment by jberkenbilt** [comment]
Based on your binary diff, it seems that something is inserting a timestamp into the header. There's an argument to tiffcmp to ignore header differences, but I don't remember what it is, and it seems to me it doesn't always work. Regarding why tiffcmp gives different results in different directions, that would probably be a bug in tiffcmp or some documented behavior. I wrote that bit of this stuff well over 10 years ago and don't recall the details. There may be ImageMagick commands to compare images too. I picked tiff for this because tiff supports multi-page images, though they are seldom used in practice, and lots of tools just ignore all but the first page. tiffcmp comes with libtiff and handles the tiff file format fully. One thing you could do is convert each page to raw samples or to ppm. ppm is basically raw samples with a trivial header that just has image dimensions. However it is possible to create a text comment at the beginning of a ppm. But most likely converting each page to ppm will result in files you can compare bitwise.

Regarding binary blobs, PDF stores binary blobs in stream objects. It would be possible to generate checksums for all the stream contents and use a matching checksum as an indicator to do a bit-wise comparison. A future version of qpdf may support detection of multiple identical streams. It is a feature that has been requested more than once. As a footnote, I will mention that it is possible to have bit differences in uncompressed streams and still have identical content in a few cases. For PDF content streams, which contain operators and so forth, whitespace is insignificant. For images, if the image width multiplied by the bits per sample is not a multiple of 8, the last few bits of the last bight of each row contains unused bits. Those bits may have any value. Also, in some cases, extraneous newlines are ignored at the ends of streams, and in some cases, trailing data after the logical end of a compressed stream is ignored. But those are unusual cases.

Also, you have me pegged when you called my response a stream of consciousness. I started by saying, "I don't have time to fully answer this right now, but here are some thoughts," and then I realized I had pretty much answered it so I went back and removed that! So many details....it's easy for me to wander down deep holes explaining all the weird corner cases. Most people don't need to know these things, but they matter when you're implementing a tool that has to handle them.

**Comment by piramiday** [comment]
> Based on your binary diff, it seems that something is inserting a timestamp into the header.

yes, that's clearly a timestamp, I don't know how I did not see that myself... sorry about that. :roll_eyes: 

> A future version of qpdf may support detection of multiple identical streams. It is a feature that has been requested more than once.

great! is there some issue/pr to subscribe to? I haven't found it by a quick search.

> Also, you have me pegged when you called my response a stream of consciousness.

let me be clear, I really appreciate the way you replied! :rocket:
I was only surprised by the scarcity of newlines. :smile: 
you were patient, clear, and to the point, nothing more can be asked of you.
unfortunately quite often folks on github will close an issue right away, e.g. as out of scope, even though they clearly have the tools to explain something to a fellow user.

so, once again, kudos to you and congrats for this awesome tool! :tophat:

**Comment by jberkenbilt** [comment]
@piramiday Thanks for your kind words. Re deduplication, you can follow #448, which I just created. The feature request was sent to me by email, but since it has come up more than once, I created the issue to track it. No guarantees on if/when I may do it.
